crypto/kms/localkeystore.go

```diff
@@ -0,0 +1,141 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kms
+
+import (
+	"sync"
+
+	"runtime"
+
+	"errors"
+
+	ec "github.com/btcsuite/btcd/btcec"
+	log "github.com/sirupsen/logrus"
+	mine "github.com/thunderdb/ThunderDB/pow/cpuminer"
+)
+
+// LocalKeyStore is the type hold local private & public key
+type LocalKeyStore struct {
+	isSet     bool
+	private   *ec.PrivateKey
+	public    *ec.PublicKey
+	nodeID    []byte
+	nodeNonce *mine.Uint256
+	sync.RWMutex
+}
+
+var (
+	// localKey is global accessible local private & public key
+	localKey     *LocalKeyStore
+	localKeyOnce sync.Once
+)
+
+var (
+	// ErrNilField indicates field is nil
+	ErrNilField = errors.New("local field is nil")
+)
+
+// InitLocalKeyStore returns a new LocalKeyStore
+func InitLocalKeyStore() {
+	localKeyOnce.Do(func() {
+		localKey = &LocalKeyStore{
+			isSet:     false,
+			private:   nil,
+			public:    nil,
+			nodeID:    nil,
+			nodeNonce: nil,
+		}
+	})
+}
+
+// SetLocalKeyPair sets private and public key, this is a one time thing
+func SetLocalKeyPair(private *ec.PrivateKey, public *ec.PublicKey) {
+	localKey.Lock()
+	defer localKey.Unlock()
+	if localKey.isSet {
+		return
+	}
+	localKey.isSet = true
+	localKey.private = private
+	localKey.public = public
+}
+
+// SetLocalNodeIDNonce sets private and public key, this is a one time thing
+func SetLocalNodeIDNonce(rawNodeID []byte, nonce *mine.Uint256) {
+	localKey.Lock()
+	defer localKey.Unlock()
+	localKey.nodeID = make([]byte, len(rawNodeID))
+	copy(localKey.nodeID, rawNodeID)
+	if nonce != nil {
+		localKey.nodeNonce = new(mine.Uint256)
+		*localKey.nodeNonce = *nonce
+	}
+}
+
+// GetLocalNodeID gets current node ID copy in []byte
+func GetLocalNodeID() (rawNodeID []byte, err error) {
+	localKey.RLock()
+	if localKey.nodeID != nil {
+		rawNodeID = make([]byte, len(localKey.nodeID))
+		copy(rawNodeID, localKey.nodeID)
+	} else {
+		err = ErrNilField
+	}
+	localKey.RUnlock()
+	return
+}
+
+// GetLocalNonce gets current node nonce copy
+func GetLocalNonce() (nonce *mine.Uint256, err error) {
+	localKey.RLock()
+	if localKey.nodeNonce != nil {
+		nonce = new(mine.Uint256)
+		*nonce = *localKey.nodeNonce
+	} else {
+		err = ErrNilField
+	}
+	localKey.RUnlock()
+	return
+}
+
+// GetLocalPublicKey gets local public key, if not set yet returns nil
+func GetLocalPublicKey() (public *ec.PublicKey, err error) {
+	localKey.RLock()
+	public = localKey.public
+	if public == nil {
+		err = ErrNilField
+	}
+	localKey.RUnlock()
+	return
+}
+
+// GetLocalPrivateKey gets local private key, if not set yet returns nil
+//  all call to this func will be logged
+func GetLocalPrivateKey() (private *ec.PrivateKey, err error) {
+	localKey.RLock()
+	private = localKey.private
+	if private == nil {
+		err = ErrNilField
```

Comment:
returning an error make passing error through callers easily, otherwise we have to write some code like this:
```golang
func Caller()(string, error) {
	someRet, err := doSomething()
	if err != nil {
		log.Errorf("something wrong: %s", err)
		return
	}

	privateKey, ok := kms.GetLocalPrivateKey()
	if !ok {
		err = errors.New("key not found")
		log.Errorf("get local private key failed: %s", err)
		return
	}
...
}
```

Reply:
Personally I prefer to do it right instead of doing it easily. But you have you point. Ok then.